### PR TITLE
timeWriteToWriter shouldn't do UTF-8 conversion.  Also rework buffering.

### PR DESCRIPTION
**Pull request by cakoose**

### CODE REVIEW

**Comment by ralfstx** [comment]
I've included your first commit (*writeToWriter should not do UTF-8 conversion*). Thank you!

As for the second commit (*Remove buffering from writeToWriter/writeToOutputStream*), I don't agree with your reasoning regarding removal of buffering in `ReadWriteBenchmark`. As you can see in `BufferedWriterBenchmark`, using a buffered writer improves performance even when writing to a `ByteArrayOutputStream`. So the buffering does not add noise to the benchmark, but it actually removes some noise by spending less time in the writer code. However, I've followed your advice to add buffering to `JsonRunner.writeToOutputStream()`.

Again, thanks a lot for your contribution!

**Comment by cakoose** [comment]
Re: my second commit.

`BufferedWriterBenchmark` writes through an `OutputStreamWriter`.  Since `OutputStreamWriter` is not efficient with small writes, wrapping a `BufferedWriter` helps there.

`ByteArrayOutputStream` is already just a memory buffer; putting another memory buffer in front of it is useless.  The same is true for wrapping `StringWriter` -- wrapping it in a `BufferedWriter` is useless.  This makes sense if you think about the underlying implementation and also holds true if you benchmark it.

My change preserved the buffering on `OutputStreamWriter` (where it matters) but remove the buffering around `ByteArrayOutputStream` and `StringWriter` (where it only adds noise).

Does this make sense?

tl;dr: remove the buffering in `timeWriteToWriter` and `timeWriteToOutputStream` to reduce noise.

**Comment by ralfstx** [comment]
Convinced!

Actually, for the `StringWriter`, the benchmark results are not so clear: while the buffering slows down most cases indeed, the minimal-json parser with the `caliper.json` input still benefits from buffering. I guess that's because there are a lot of small values in this input leading to very many single char writes. And writing to the `BufferedWriter` seems to be slightly more efficient than to the `StringBuffer` that is contained in the `StringWriter`. It seems that I should implement a small write cache to improve writing to plain `StringWriters`, which would also improve writing Strings.

**Comment by cakoose** [comment]
Oh, interesting.  I didn't actually benchmark the `BufferedWriter` + `StringWriter` case.

Looking at OpenJDK 7 source code, it looks like `StringWriter` uses `StringBuffer` underneath, which preallocates 16 chars of space by default.  `BufferedWriter` preallocates 8192 chars of space by default.  So I guess the buffer resize cost is outweighing the cost of the extra copy.

I've seen other serializer benchmarks that pre-allocate the target buffer to the exact correct size (i.e. run the serializer once, see how big the output is, then preallocate the `StringWriter` to exactly that size).  For the purposes of this kind of benchmark, array resize operations are arguably just noise.  This could be a way to get rid of that noise.

**Comment by ralfstx** [comment]
I've pre-allocated the buffers as you suggested. It doesn't seem to make a huge difference, but it also won't do any harm. But still, a writing buffer speeds up the JsonWriter, so it's not just the buffer resize cost.

I've now added a small (128 chars) writing buffer to minimal-json that leads to even better results than a `BufferedWriter`. It may not have the same effect for all types of Writers, but at least it decouples the writing performance from the performance of the underlying writer without adding much overhead and also speeds up `toString()`. Writing to a `Writer` is now just as fast as Jackson (even slightly faster for the caliper.json).
